File: Day3.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TupleSections #-}

import qualified Data.Attoparsec.Text as At
import qualified Data.Attoparsec.Combinator as Atc
import qualified Data.Text as T
import Control.Applicative
import Control.Monad
import Data.Bifunctor (first, second, bimap)
import Data.Functor
import Data.Array
import Data.Char (digitToInt)
import qualified Data.Set as S

import AocLib

data Cell = Empty | Symbol | Gear | Value {value :: Int} deriving (Show, Eq)
type GIx = (Int, Int)
type Grid = Array GIx Cell

isValueCell :: Cell -> Bool
isValueCell (Value _) = True
isValueCell _ = False

parser :: At.Parser Grid
parser = do
    let parseCell = At.choice [Value . digitToInt <$> At.digit,
                               At.char '.' $> Empty,
                               At.char '*' $> Gear,
                               At.notChar '\n' $> Symbol]
    cells <- At.sepBy1 (At.many1 parseCell) At.endOfLine
    let rows = length cells
        cols = length $ head cells
    return $ listArray ((0, 0), (rows - 1, cols - 1)) (concat cells)

type NumPosition = (Int, S.Set GIx)

extractNums :: Grid -> [NumPosition]
extractNums grid = do
    (ir, ic) <- range $ bounds grid
    let ix = (ir, ic)
    guard (isValueCell $ grid ! ix)
    let leftix = (ir, ic - 1)
        validIx = inRange $ bounds grid
    guard (not (validIx leftix) || not (isValueCell (grid ! leftix)))
    let rightixs = takeWhile (\it -> validIx it && isValueCell (grid ! it)) $ map (ir, ) [ic+1..]
        digits = map (value . (grid !)) (ix : rightixs)
        asNum = foldl1 ((+) . (* 10)) digits
    return (asNum, S.fromList (ix : rightixs))

solve1 :: [NumPosition] -> Grid -> Int
solve1 nums grid = sum $ do
    (n, ixs) <- nums
    let isValidIx = inRange (bounds grid)
        left = second pred $ S.findMin ixs
        right = second succ $ S.findMax ixs
        up = map (first pred) (left : right : S.toList ixs)
        down = map (first succ) (left : right : S.toList ixs)
    guard (any ((== Symbol) . (grid !)) (filter isValidIx ([left, right] ++ up ++ down)))
    return n

solve2 :: [NumPosition] -> Grid -> Int
solve2 nums grid = sum $ do
    let bs = bounds grid
    ix@(ir, ic) <- range bs
    guard (grid ! ix == Gear)
    let dixs = S.fromList $ filter (inRange bs) [bimap (+ di) (+ dj) ix | di <- [-1..1], dj <- [-1..1], (di, dj) /= (0, 0)]
        matchingNums = filter (not . S.null . S.intersection dixs . snd) nums
    guard (length matchingNums == 2)
    return $ product (map fst matchingNums)

main :: IO ()
main = do
    input <- parseInput parser
    let numbers = extractNums input
    print (solve1 numbers input)
    print (solve2 numbers input)

File: day_3.rs
use std::error::Error;
use aoclib_rs::aoc::*;

use nom::*;

#[derive(Eq, PartialEq, Debug, Copy, Clone)]
enum Cell {
    Empty,
    Symbol,
    Gear,
    Value(i32),
}

type Ix = Ixn<2>;

#[derive(Debug)]
struct Grid {
    nums: Vec<(Cell, Vec<Ix>)>,
    dim: i32,
    cells: std::collections::BTreeMap<Ix, Cell>,
}

fn ixf(el: [i32; 2]) -> Ix { Ixn::<2>(el) }

fn parse(i: &str) -> Parsed<Grid> {
    fn cell(j: &str) -> Parsed<Cell> {
        let digit = combinator::map(
            character::complete::satisfy(|c| c.is_ascii_digit()),
            |c: char| c.to_digit(10).unwrap());

        let symbol = character::complete::satisfy(|c| c.is_ascii_graphic());

        branch::alt((
            combinator::value(Cell::Empty, character::complete::char('.')),
            combinator::value(Cell::Gear, character::complete::char('*')),
            combinator::map(digit, |v| Cell::Value(v as i32)),
            combinator::value(Cell::Symbol, symbol),
        ))(j)
    }

    fn line(j: &str) -> Parsed<Vec<(usize, Cell)>> {
        combinator::map(multi::many1(cell),
                        |vs| vs.iter().copied().enumerate().collect())(j)
    }

    fn coalesce(vs: Vec<(usize, Cell)>) -> Vec<(Cell, Vec<i32>)> {
        let mut rs = Vec::<(Cell, Vec<i32>)>::new();
        for (j, cell) in vs {
            match cell {
                Cell::Value(t) => match rs.last_mut() {
                    Some((Cell::Value(ref mut n), ref mut ixs)) => {
                        ixs.push(j as i32);
                        *n = *n * 10 + t;
                    }
                    _ => rs.push((cell, vec![j as i32]))
                }
                _ => rs.push((cell, vec![j as i32]))
            }
        }
        rs
    }

    let (i, lines) = multi::separated_list1(character::complete::newline,
                                            combinator::map(line, coalesce))(i)?;

    let nums: Vec<(Cell, Vec<Ix>)> = lines.iter()
        .enumerate()
        .flat_map(|(i, row)|
            row.iter()
                .map(move |(c, js)|
                    (*c, std::iter::repeat(i as i32).zip(js).map(|(a, b)| Ixn([a, *b])).collect())))
        .filter(|(v, _)| *v != Cell::Empty).collect();
    let dim = lines.len() as i32;
    let cells = nums.iter()
        .flat_map(|(c, ixs)|
            ixs.iter().copied().zip(std::iter::repeat(*c))).collect();
    Ok((i, Grid { dim, nums, cells }))
}

fn solve1(Grid { dim, nums, cells }: &Grid) -> i32 {
    let valid_ix = |ix: &Ix| ix.in_range(&Ix::from(0), &Ix::from(dim - 1));

    let has_gear = |ix: &Ix| valid_ix(ix)
        && cells.get(ix).map(|v| *v == Cell::Symbol || *v == Cell::Gear)
        .unwrap_or(false);

    nums.iter()
        .filter(|(c, _)| matches!(c, Cell::Value(_)))
        .filter(|(_, ixs)| {
            let left: Ix = ixf([0, -1]) + *ixs.first().unwrap();
            let right: Ix = ixf([0, 1]) + *ixs.last().unwrap();
            let sides = [left, right];
            let up = ixs.iter().copied().chain(sides).map(|ix| ixf([-1, 0]) + ix);
            let down = ixs.iter().copied().chain(sides).map(|ix| ixf([1, 0]) + ix);
            up.chain(down).chain([left, right]).into_iter().any(|e| has_gear(&e))
        })
        .map(|e| if let Cell::Value(n) = e.0 { n } else { 0 })
        .sum()
}

fn solve2(Grid { dim: _, nums, cells }: &Grid) -> i32 {
    nums.iter()
        .filter_map(|(c, ixs)| if matches!(c, Cell::Gear) { Some(ixs[0]) } else { None })
        .filter_map(|ix| {
            let deltas = |x: &Ix| (-1..2)
                .flat_map(|i| (-1..2)
                    .map(move |j| ixf([i, j])))
                .filter(|ix| *ix != Ix::from(0))
                .map(|r| *x + r)
                .filter(|ix| cells.contains_key(ix)).collect::<Vec<_>>();

            let nums_around: std::collections::HashSet<i32> = deltas(&ix).iter()
                .filter_map(|i| match cells.get(&i) {
                    Some(Cell::Value(n)) => Some(*n),
                    _ => None,
                }).collect();
            if nums_around.len() == 2 { Some(nums_around.iter().product::<i32>()) } else { None }
        })
        .sum()
}

fn main() -> Result<(), Box<dyn Error>> {
    let input = parse_input(parse)?;
    println!("{}", solve1(&input));
    println!("{}", solve2(&input));
    Ok(())
}
